File: packages/mapping/src/distance.ts
const IDLE_RESET_MS = 1000;
const MIN_SPAN = 0.02;

export class DistanceCalibrator {
  private autoMin: number | null = null;
  private autoMax: number | null = null;
  private lastCalibrateMs = 0;

  constructor(
    private cfgMin: number | null = null,
    private cfgMax: number | null = null
  ) {}

  map(raw: number, calibrating: boolean, timestampMs: number): number {
    if (calibrating) {
      this.lastCalibrateMs = timestampMs;
      this.updateAuto(raw);
    } else if (timestampMs - this.lastCalibrateMs > IDLE_RESET_MS) {
      this.autoMin = null;
      this.autoMax = null;
    }

    const range = this.range();
    if (range.lo == null || range.hi == null) {
      return Math.max(0, Math.min(1, raw));
    }
    const span = Math.max(range.hi - range.lo, MIN_SPAN);
    return Math.max(0, Math.min(1, (raw - range.lo) / span));
  }

  private range(): { lo: number | null; hi: number | null } {
    if (this.autoMin != null && this.autoMax != null) {
      if (this.autoMax - this.autoMin >= MIN_SPAN) {
        return { lo: this.autoMin, hi: this.autoMax };
      }
    }
    if (this.cfgMin != null && this.cfgMax != null) {
      return { lo: this.cfgMin, hi: this.cfgMax };
    }
    return { lo: null, hi: null };
  }

  private updateAuto(raw: number): void {
    if (this.autoMin == null) {
      this.autoMin = raw;
      this.autoMax = raw;
      return;
    }
    if (this.autoMin != null && raw < this.autoMin) {
      this.autoMin = this.autoMin * 0.88 + raw * 0.12;
    } else if (this.autoMax != null && raw > this.autoMax) {
      this.autoMax = this.autoMax * 0.88 + raw * 0.12;
    }
  }
}

File: src/expression/distance.py
"""Calibração de distância entre mãos → volume 0–1."""

from __future__ import annotations

_IDLE_RESET_MS = 1000
_MIN_SPAN = 0.04


class DistanceCalibrator:
    """Auto-calibra por sessão; limites YAML só como fallback inicial."""

    def __init__(
        self,
        dist_min: float | None = None,
        dist_max: float | None = None,
    ) -> None:
        self._cfg_min = dist_min
        self._cfg_max = dist_max
        self._auto_min: float | None = None
        self._auto_max: float | None = None
        self._last_calibrate_ms = 0

    def map(self, raw: float, calibrating: bool, timestamp_ms: int) -> float:
        if calibrating:
            self._last_calibrate_ms = timestamp_ms
            self._update_auto(raw)
        elif timestamp_ms - self._last_calibrate_ms > _IDLE_RESET_MS:
            self._auto_min = None
            self._auto_max = None

        lo, hi = self._range()
        if lo is None or hi is None:
            return max(0.0, min(1.0, raw))

        span = max(hi - lo, _MIN_SPAN)
        return max(0.0, min(1.0, (raw - lo) / span))

    def _range(self) -> tuple[float | None, float | None]:
        if self._auto_min is not None and self._auto_max is not None:
            if self._auto_max - self._auto_min >= _MIN_SPAN:
                return self._auto_min, self._auto_max
        if self._cfg_min is not None and self._cfg_max is not None:
            return self._cfg_min, self._cfg_max
        return None, None

    def _update_auto(self, raw: float) -> None:
        if self._auto_min is None:
            self._auto_min = raw
            self._auto_max = raw
            return
        if raw < self._auto_min:
            self._auto_min = self._auto_min * 0.88 + raw * 0.12
        elif raw > self._auto_max:
            self._auto_max = self._auto_max * 0.88 + raw * 0.12
